src/elastio-snap.c

```diff
@@ -2271,13 +2537,173 @@ static int cow_sync_and_close(struct cow_manager *cm){
 	return ret;
 }
 
+static inline void elastio_snap_mm_lock(struct mm_struct *mm)
+{
+	// TODO: fix for other kernel versions
+	down_write(&mm->mmap_sem);
+}
+
+static inline void elastio_snap_mm_unlock(struct mm_struct *mm)
+{
+	// TODO: fix for other kernel versions
+	up_write(&mm->mmap_sem);
+}
+
+static int elastio_snap_get_cow_file_extents(struct snap_device *dev, struct file *filp)
+{
+	int ret;
+	struct fiemap_extent_info fiemap_info;
+	unsigned int fiemap_mapped_extents_size, i_ext;
+	struct fiemap_extent *extent;
+	char parent_process_name[TASK_COMM_LEN];
+	unsigned long vm_flags = VM_READ | VM_WRITE;
+	unsigned long start_addr;
+	struct task_struct *task;
+	struct vm_area_struct *vma;
+	struct page *pg;
+	__user uint8_t *cow_ext_buf;
+	size_t cow_ext_buf_size;
+
+	int (*fiemap)(struct inode *, struct fiemap_extent_info *, u64 start, u64 len);
+
+	struct vm_area_struct * (*vm_area_alloc)(struct mm_struct *mm) = (VM_AREA_ALLOC_ADDR != 0) ?
+        (struct vm_area_struct * (*)(struct mm_struct *mm)) (VM_AREA_ALLOC_ADDR + (long long)(((void *)kfree) - (void *)KFREE_ADDR)) : NULL;
+
+	void (*vm_area_free)(struct vm_area_struct *vma) = (VM_AREA_FREE_ADDR != 0) ?
+        (void (*)(struct vm_area_struct *vma)) (VM_AREA_FREE_ADDR + (long long)(((void *)kfree) - (void *)KFREE_ADDR)) : NULL;
+
+	int (*insert_vm_struct)(struct mm_struct *mm, struct vm_area_struct *vma) = (INSERT_VM_STRUCT_ADDR != 0) ?
+        (int (*)(struct mm_struct *mm, struct vm_area_struct *vma)) (INSERT_VM_STRUCT_ADDR + (long long)(((void *)kfree) - (void *)KFREE_ADDR)) : NULL;
+
+	if (!vm_area_alloc) {
+		LOG_ERROR(-ENOTSUPP, "vm_area_alloc() was not found");
+		return -ENOTSUPP;
+	}
+
+	if (!vm_area_free) {
+		LOG_ERROR(-ENOTSUPP, "vm_area_free() was not found");
+		return -ENOTSUPP;
+	}
+
+	if (!insert_vm_struct) {
+		LOG_ERROR(-ENOTSUPP, "insert_vm_struct() was not found");
+		return -ENOTSUPP;
+	}
```

Comment:
Are you sure we have to do this address magic in the function each time when it called instead of static option?

Reply:
Perhaps. But we'll need to check for nulls anyway

Reply:
Sure we'll need to check. I mean the addresses don't change at runtime. And it can be somewhere outside of the function like this
https://github.com/elastio/elastio-snap/blob/f64080159e5c5f0829a4586dccebcce6340c8694/src/elastio-snap.c#L144-L145

Reply:
In that case we'll try to redefine an existing function (f.e., `vm_area_alloc()`) existing in the linux header but which isn't exported.

Reply:
But wouldn't be the same when `elastio_snap_get_cow_file_extents` is called? This can be prevented in other ways. Let's talk about it later.